ussologin/ussologin.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package ussologin defines functionality used for allowing clients
+// to authenticate with the IDM server using USSO OAuth.
+package ussologin
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/httprequest"
+	"github.com/juju/usso"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/environschema.v1"
+	"gopkg.in/juju/environschema.v1/form"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+type tokenGetter interface {
+	GetTokenWithOTP(username, password, otp, tokenName string) (*usso.SSOData, error)
+}
+
+// This is defined here to allow it to be stubbed out in tests
+var server tokenGetter = usso.ProductionUbuntuSSOServer
+
+// Login completes the login information using the provided filler
+// and attempts to obtain a USSO token using this information.
+// If the store is non-nil it is used to store this token.
+func Login(filler form.Filler) (*usso.SSOData, error) {
+	login, err := filler.Fill(loginForm)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot read login parameters")
+	}
+	tok, err := server.GetTokenWithOTP(
+		login["username"].(string),
+		login["password"].(string),
+		login["otp"].(string),
+		"charm",
+	)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot get token")
+	}
+	return tok, nil
+}
+
+var loginForm = form.Form{
+	Fields: environschema.Fields{
+		"username": environschema.Attr{
+			Description: "Username",
+			Type:        environschema.Tstring,
+			Mandatory:   true,
+			Group:       "1",
+		},
+		"password": environschema.Attr{
+			Description: "Password",
+			Type:        environschema.Tstring,
+			Mandatory:   true,
+			Secret:      true,
+			Group:       "1",
+		},
+		"otp": environschema.Attr{
+			Description: "Two-factor auth",
+			Type:        environschema.Tstring,
+			Group:       "2",
+		},
+	},
+}
+
+// doSignedRequest performs a GET request to the given URL signed using the provided token.
+func doSignedRequest(client *http.Client, ussoAuthUrl string, tok *usso.SSOData, u *url.URL) error {
```

Comment:
I think we'll find we want this exported, but we can always do that when we need it.